pulse2percept/retina.py

```diff
@@ -1,772 +1,958 @@
+        # means that response magnitudes don't change as you increase the
+        # length of axonal integration or sampling of the retina
+        # Doesn't affect normalization over time, or responses as a function
+        # of the anount of current,
+
+        return ecs
+
+    def electrode_ecs(self, implant, alpha=14000, n=1.69):
+        """
+        Gather current spread and effective current spread for each electrode
+        within both the bipolar and the ganglion cell layer
+
+        Parameters
+        ----------
+        implant : implants.ElectrodeArray
+            An implants.ElectrodeArray instance describing the implant.
+
+        alpha : float
+            Current spread parameter
+        n : float
+            Current spread parameter
+
+        Returns
+        -------
+        ecs : contains n arrays containing the the effective current
+            spread within various layers
+            for each electrode in the array respectively.
+
+        See also
+        --------
+        Electrode.current_spread
+        """
+
+        cs = np.zeros((self.gridx.shape[0], self.gridx.shape[1],
+                       2, len(implant.electrodes)))
+        ecs = np.zeros((self.gridx.shape[0], self.gridx.shape[1],
+                        2, len(implant.electrodes)))
+
+        for i, e in enumerate(implant.electrodes):
+            cs[..., 0, i] = e.current_spread(self.gridx, self.gridy,
+                                             layer='INL', alpha=alpha, n=n)
+            ecs[..., 0, i] = cs[..., 0, i]
+            cs[..., 1, i] = e.current_spread(self.gridx, self.gridy,
+                                             layer='OFL', alpha=alpha, n=n)
+            ecs[:, :, 1, i] = self.current2effectivecurrent(cs[..., 1, i])
+
+        return ecs, cs
+
+
+@six.add_metaclass(abc.ABCMeta)
+class TemporalModel():
+    """Abstract base class for all models of temporal sensitivity.
+
+    This class provides a standard template for all models of temporal
+    sensitivity.
+    """
+
+    def set_kwargs(self, warn_inexistent, **kwargs):
+        """Overwrite any given keyword arguments
+
+        Parameters
+        ----------
+        warn_inexistent : bool
+            If True, displays a warning message if a keyword is provided that
+            is not recognized by the temporal model.
+        """
+        for key, value in kwargs.items():
+            if not hasattr(self, key) and warn_inexistent:
+                w_s = "Unknown class attribute '%s'" % key
+                logging.getLogger(__name__).warning(w_s)
+            setattr(self, key, value)
+
+    def __init__(self, **kwargs):
+        self.set_kwargs(True, **kwargs)
+
+    @abc.abstractmethod
+    def model_cascade(self, in_arr, pt_list, layers, dojit):
+        """Custom ganglion cell model
+
+        Parameters
+        ----------
+        in_arr: array-like
+            A 2D array specifying the effective current values
```

Comment:
Not a big deal: is it just me, or are these lines all wrapped to less than 80 characters?